### ISSUE TITLE
Enhancement: make it easier to compile jparse

### ISSUE THREAD

**Issue by lcn2** [xexyl/jparse#2]
### Is there an existing issue for this?

- [X] I have searched for existing issues and did not find anything like this

### Describe the feature

In order to compile jparse, we had to clone the code under a sub-directed we called `njparse/jparse` and then setup these symlinks under `njparse`:

```
dbg -> ../mkiocccentry/dbg
dyn_array -> ../mkiocccentry/dyn_array
iocccsize.h -> ../mkiocccentry/iocccsize.h
```

### Relevant images, screenshots or other files

```
$ ls -l
total 0
lrwxr-xr-x  1 chongo   19 Aug 28 13:37 dbg -> ../mkiocccentry/dbg
lrwxr-xr-x  1 chongo   25 Aug 28 13:37 dyn_array -> ../mkiocccentry/dyn_array
lrwxr-xr-x  1 chongo   27 Aug 28 13:37 iocccsize.h -> ../mkiocccentry/iocccsize.h
drwxr-xr-x 42 chongo 1344 Aug 28 13:52 jparse

$ cd jparse
$ git remote -v
origin	git@github.com:lcn2/jparse.git (fetch)
origin	git@github.com:lcn2/jparse.git (push)

$ cd ../../mkiocccentry
$ git remote -v
origin	git@github.com:ioccc-src/mkiocccentry.git (fetch)
origin	git@github.com:ioccc-src/mkiocccentry.git (push)
```

### Relevant links

- https://github.com/ioccc-src/mkiocccentry
- https://github.com/lcn2/dbg
- https://github.com/lcn2/dyn_array

### Anything else?

We recommend that this repo assume that the [dbg repo be installed](https://github.com/lcn2/dbg) **AND** that the [dyn_array repo be installed](https://github.com/lcn2/dyn_array).

People who wish to compile and use this code must assume that various prerequisites be available.  Instructions in this repo's `README.md` can let then know how to do this.

### Final TODO

- [x] Do the tasks in [comment 2322179729](https://github.com/xexyl/jparse/issues/2#issuecomment-2322179729) to make sure everything is okay in the mkiocccentry repo.

**Comment by xexyl** [comment]
I agree with you on that. I just put a note in the README.md about populating the repo. But you bring up an interesting point - that it could be that we require them to install the repo.

Perhaps the workflow could be updated to clone those repos.

**Comment by lcn2** [comment]
BTW: We are in the process of updating [](https://github.com/lcn2/dbg) and [](https://github.com/lcn2/dyn_array) as per the contents of the mkiocccentry code.

**Comment by xexyl** [comment]
> BTW: We are in the process of updating and as per the contents of the mkiocccentry code.

Okay in that case I will not address this now. Please advise (or open a pull request) when you have it figured out. In the meantime I will be taking care of something else.

Thanks!

**Comment by lcn2** [comment]
We have updated:

- https://github.com/lcn2/dbg
- https://github.com/lcn2/dyn_array

They now are copies of the mkiocccentry tree as per tag v1.5.1

**Comment by xexyl** [comment]
> We have updated:
> 
> * https://github.com/lcn2/dbg
> * https://github.com/lcn2/dyn_array
> 
> They now are copies of the mkiocccentry tree as per tag v1.5.1

Thanks. Now the question is what will be easier for you for future contests (if say needing to update the code here):

0. Have the libraries part of this repo like with mkiocccentry.
1. Tell users where to get the libraries and how to install them (and of course updating the workflow to clone and make and install those repos).

Either way works though I think it would be nice to have it 'just work'. Nonetheless if it's easier to not have it that way then I can do that.

**Comment by lcn2** [comment]
There is the problem of `iocccsize.h`.  It is included by `jsemtblgen.h` and obviously should not.

We don't have a solution for that just yet.

**Comment by xexyl** [comment]
> There is the problem of `iocccsize.h`. It is included by `jsemtblgen.h` and obviously should not.
> 
> We don't have a solution for that just yet.

Well I already removed that .. at least here, since it's not needed (or it seems to not be).

Or do you mean something else?

**Comment by xexyl** [comment]
Hmm ... yes I see what you mean with the symlinks. The reason it worked for me is NOT because it was installed. Of course not. It's because I have those subdirectories in the same directory.

A problem is that if we remove the `../` from the includes then in mkiocccentry it will break. But if we don't it means the users of this repo will have to have those libraries in that precise location. Also if the subdirectories for `dbg` etc. were in `jparse/` then it would require a change in how mkiocccentry is structured.

So a number of things have to be decided on that too.

**Comment by lcn2** [comment]
> 1. Tell users where to get the libraries and how to install them (and of course updating the workflow to clone and make and install those repos).

This is probably best in the long run.

There is nothing unusual about a repo depending on code from another repo via installed libraries, installed header files, and installed man pages.

**Comment by lcn2** [comment]
We removed the dependence in the mkiocccentry repo with commit a86c687478d63f42fa46b9f64c1dbe5f9ca1ffe5

**Comment by xexyl** [comment]
> > 1. Tell users where to get the libraries and how to install them (and of course updating the workflow to clone and make and install those repos).
> 
> This is probably best in the long run.
> 
> There is nothing unusual about a repo depending on code from another repo via installed libraries, installed header files, and installed man pages.

That is true but the problem is the way it's included, the `../`. If we do the usual include (no directory) then we also require that users of the mkiocccentry toolkit have to install those libraries. But if we don't do that then we have the problem of how to clone this repo to the mkiocccentry repo? It would be complicated I think, since every time it would have to be edited (at least that is so based on how I'm thinking - or not - about it right now).

**Comment by lcn2** [comment]
BTW: We do not need to update `mkiocccentry/jparse` all the time.  However we do recommend that once this issue is resolved, that we do so for mkiocccentry v1.5.2 (that as just started).

**Comment by xexyl** [comment]
> BTW: We do not need to update `mkiocccentry/jparse` all the time. However we do recommend that once this issue is resolved, that we do so for mkiocccentry v1.5.2 (that as just started).

Sure. But do we need users to install the libraries?

If not the paths here and in the mkiocccentry repo will differ which obviously won't work. If yes then we need to update the paths over there, too, and instructions and probably Makefiles also. But it seems like we need this code in the other repo.

Either way the issue of paths of include files has to be resolved in a way that works for both this repo and the other one. I think forcing users to have to have directories of `../dbg` etc. is not very kind. But on the other hand having a copy of the libraries in this repo and in mkiocccentry is also not that great of a solution. I am fine with what you prefer but I personally think that if we require users to have a copy of the other libraries that it would be installed, assuming it can't be in the repo (but if it was in this repo there would be a path level difference here than in mkiocccentry which means that the `../` will not work the same way).

**Comment by lcn2** [comment]
We recommend that this repo assume that the prerequisites be installed.

The [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) is a special case where IOCCC submitters needed a complete package.

Perhaps the use of some `#if defined(MKIOCCCENTRY_SRC)` could be used?

In the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) the upper level `Makefile` would force it to be defined.  But in this repo, it would not be defined by default.

So we would have:

```c
/*
 * dbg - info, debug, warning, error, and usage message facility
 */
#if defined(MKIOCCCENTRY_SRC)
#include "../dbg/dbg.h"
#else
#include </dbg/dbg.h>
#endif
```

For example.

**Comment by xexyl** [comment]
> We recommend that this repo assume that the prerequisites be installed.
> 
> The [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) is a special case where IOCCC submitters needed a complete package.
> 
> Perhaps the use of some `#if defined(MKIOCCCENTRY_SRC)` could be used?
> 
> In the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) the upper level `Makefile` would force it to be defined. But in this repo, it would not be defined by default.
> 
> So we would have:
> 
> ```c
> /*
>  * dbg - info, debug, warning, error, and usage message facility
>  */
> #if defined(MKIOCCCENTRY_SRC)
> #include "../dbg/dbg.h"
> #else
> #include </dbg/dbg.h>
> #endif
> ```
> 
> For example.

Perhaps you'd like to submit a pull request on that? Because that does seem like a good idea. When this is done I can update the workflows to install those libraries, so that your suggestion works.

**Comment by lcn2** [comment]
> > We recommend that this repo assume that the prerequisites be installed.
> > The [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) is a special case where IOCCC submitters needed a complete package.
> > Perhaps the use of some `#if defined(MKIOCCCENTRY_SRC)` could be used?
> > In the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) the upper level `Makefile` would force it to be defined. But in this repo, it would not be defined by default.
> > So we would have:
> > ```c
> > /*
> >  * dbg - info, debug, warning, error, and usage message facility
> >  */
> > #if defined(MKIOCCCENTRY_SRC)
> > #include "../dbg/dbg.h"
> > #else
> > #include </dbg/dbg.h>
> > #endif
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > For example.
> 
> Perhaps you'd like to submit a pull request on that? Because that does seem like a good idea. When this is done I can update the workflows to install those libraries, so that your suggestion works.

In this repo or the other repo?

**Comment by xexyl** [comment]
> > > We recommend that this repo assume that the prerequisites be installed.
> > > The [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) is a special case where IOCCC submitters needed a complete package.
> > > Perhaps the use of some `#if defined(MKIOCCCENTRY_SRC)` could be used?
> > > In the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) the upper level `Makefile` would force it to be defined. But in this repo, it would not be defined by default.
> > > So we would have:
> > > ```c
> > > /*
> > >  * dbg - info, debug, warning, error, and usage message facility
> > >  */
> > > #if defined(MKIOCCCENTRY_SRC)
> > > #include "../dbg/dbg.h"
> > > #else
> > > #include </dbg/dbg.h>
> > > #endif
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > For example.
> > 
> > 
> > Perhaps you'd like to submit a pull request on that? Because that does seem like a good idea. When this is done I can update the workflows to install those libraries, so that your suggestion works.
> 
> In this repo or the other repo?

This repo. But I just checked and did not see that define in the other repo so that obviously has to be added too (but where I don't know). I think it would work though, the way you suggested.

**Comment by lcn2** [comment]
> Sure. But do we need users to install the libraries?

Should the mkiocccentry Makefile just use `-ldgb` (instead of `../dbg/dbg.a`) and let the mkiocccentry Makefile control the path from which the `dbg.a` file is found?

Not sure, but something like that might be doable.

**Comment by xexyl** [comment]
> > Sure. But do we need users to install the libraries?
> 
> Should the mkiocccentry Makefile just use `-ldgb` (instead of `../dbg/dbg.a`) and let the mkiocccentry Makefile control the path from which the `dbg.a` file is found?
> 
> Not sure, but something like that might be doable.

On the one hand it might make it easier for synchronising this repo with that one .. but on the other hand it would require users to install the other repos. Now that might not be a problem (if it is something they should do then obviously they will have to install this one too) but if it is done it will (I think, in my tired head) require a number of changes there too, some perhaps significant (like removal of those subdirectories and changing the paths for include files?).

**Comment by xexyl** [comment]
> > > Sure. But do we need users to install the libraries?
> > 
> > 
> > Should the mkiocccentry Makefile just use `-ldgb` (instead of `../dbg/dbg.a`) and let the mkiocccentry Makefile control the path from which the `dbg.a` file is found?
> > Not sure, but something like that might be doable.
> 
> On the one hand it might make it easier for synchronising this repo with that one .. but on the other hand it would require users to install the other repos. Now that might not be a problem (if it is something they should do then obviously they will have to install this one too) but if it is done it will (I think, in my tired head) require a number of changes there too, some perhaps significant (like removal of those subdirectories and changing the paths for include files?).

The question is what is preferable for the IOCCC, I think, as it's the primary use of this repo. I recommend that if we require users to install these things we can provide a set of commands that they might use to install everything (a script maybe but don't know).

**Comment by xexyl** [comment]
> As part of [GH-issuecomment-2316314392)](https://github.com/xexyl/jparse/issues/2#issuecomment-2316314392) we had to update the [dyn_array repo](https://github.com/lcn2/dyn_array).
> 
> You will need to update and re-install from the [dyn_array repo](https://github.com/lcn2/dyn_array).
> 
> See PR #3 for the include file fix.

How does this affect mkiocccentry too? And thanks! When you're done with this stuff I'll update the README.md file and also update the workflows.

**Comment by xexyl** [comment]
Running make depend ...

**Comment by lcn2** [comment]
> > As part of [GH-issuecomment-2316314392)](https://github.com/xexyl/jparse/issues/2#issuecomment-2316314392) we had to update the [dyn_array repo](https://github.com/lcn2/dyn_array).
> > You will need to update and re-install from the [dyn_array repo](https://github.com/lcn2/dyn_array).
> > See PR #3 for the include file fix.
> 
> How does this affect mkiocccentry too? And thanks! When you're done with this stuff I'll update the README.md file and also update the workflows.

Yes, but we suggest that PR #3 be processed and then we will update the mkiocccentry repo accordingly including forcing `MKIOCCCENTRY_SRC` to be defined there by default.

**Comment by xexyl** [comment]
> > > As part of [GH-issuecomment-2316314392)](https://github.com/xexyl/jparse/issues/2#issuecomment-2316314392) we had to update the [dyn_array repo](https://github.com/lcn2/dyn_array).
> > > You will need to update and re-install from the [dyn_array repo](https://github.com/lcn2/dyn_array).
> > > See PR #3 for the include file fix.
> > 
> > 
> > How does this affect mkiocccentry too? And thanks! When you're done with this stuff I'll update the README.md file and also update the workflows.
> 
> Yes, but we suggest that PR #3 be processed and then we will update the mkiocccentry repo accordingly including forcing `MKIOCCCENTRY_SRC` to be defined there by default.

Already did, thanks!

**Comment by xexyl** [comment]
Just updated the workflow. Hopefully that solves the problems. Once you've done the other part then (likely tomorrow) I can update the README.md file here. Thanks for the help! I think it's good to get this in early actually.

**Comment by xexyl** [comment]
Looks like dyn_array needs to have a make depend update?

**Comment by xexyl** [comment]
Not only that but a problem with make depend in macOS. With `-MM` it should NOT include system paths but this was done with `make depend`.

It uses your tool independ but I am not sure if that's a problem with that tool or not. Please advise. I wish I had seen this before the commit but I guess it helped uncover another problem.

**Comment by xexyl** [comment]
Hmm .. looks like it might be because it does not use `cc -MM`? That's a problem obviously.

**Comment by lcn2** [comment]
The mkiocccentry repo is in now sync with that repo's commit c51ae6a1660c262d07361aded73afceaa20c8b73

**Comment by xexyl** [comment]
> The mkiocccentry repo is in now sync with that repo's commit c51ae6a1660c262d07361aded73afceaa20c8b73

Thanks. I just updated the README.md file again .. a minor fix.

It occurred to me that once this is all done the IOCCC guidelines and maybe rules (not sure if both need it) have to be updated too.

And as above there are some issues with some of the other repos too that have to be resolved before this can be complete. I will likely be leaving soon for the day but hopefully tomorrow we can finish this off! I did fix a few more entries (html files) in the website btw.

**Comment by lcn2** [comment]
Yes, the code in mkiocccentry is now broken on Linux.

We have to go now .. perhaps you have a fix for that repo?

**Comment by xexyl** [comment]
> Yes, the code in mkiocccentry is now broken on Linux.
> 
> We have to go now .. perhaps you have a fix for that repo?

I can look. I will probably have to go soon too though.

**Comment by xexyl** [comment]
I made an update to hostchk.sh which skips over `<dbg.h>` and `<dyn_array.h>` but unfortunately there is still another issue with the Makefile dependencies. That is your tool making it a system path which isn't always the case. I think you'll have to address this one but the hostchk.sh update works past the new includes.

**Comment by xexyl** [comment]
Although I mentioned it already I'm mentioning it again as a reminder (as I'm more likely to see it this way). Once this issue is resolved and people can use this repo on their own the guidelines and maybe rules have to be updated too. I'll take care of that. I would do that now except that there are some issues to be resolved still like with the make dependencies (this problem also exists in mkiocccentry as noted but it's also here unfortunately).

I am shutting down now. Hopefully tomorrow we can get everything sorted. That would be great!

**Comment by lcn2** [comment]
With commit 1434b45aba52d32e593a1963d56a15376e403543 we have a fix for `test_ioccc/hostchk.sh`, but more is needed.

**Comment by xexyl** [comment]
It might be nice if we could make use of the system used in mkiocccentry so that the workflows can pass and so that people can more easily compile this repo. For instance it might be something like (I am typing this with bleary eyes btw):

```c
#ifdef JPARSE_REPO
#include "dbg/dbg.h"
/* etc. */
#elif defined (MKIOCCCENTRY_SRC)
#include "../dbg/dbg.h"
/* etc. */
#endif 
```

This would not only make it easy for the workflow to work right but it would make it easier for the users of the repo to use the repo. True it is not unreasonable to make them install the other repos but it would also be nice to have the workflow pass here.

I am not sure if you had to make changes to the make depend script or not. It seems not and I think having in the make dependencies the full path of include files is not ideal but of course the above idea would also make this not be a problem (at least as I think in my tired head).

Hope you're having a nice sleep. Unsure if I'll play with the idea in this comment or not (yet?).

**Comment by lcn2** [comment]
> It might be nice if we could make use of the system used in mkiocccentry so that the workflows can pass and so that people can more easily compile this repo. For instance it might be something like (I am typing this with bleary eyes btw):
> 
> 
> 
> ```c
> 
> #ifdef JPARSE_REPO
> 
> #include "dbg/dbg.h"
> 
> /* etc. */
> 
> #elif defined (MKIOCCCENTRY_SRC)
> 
> #include "../dbg/dbg.h"
> 
> /* etc. */
> 
> #endif 
> 
> ```
> 
> 
> 
> This would not only make it easy for the workflow to work right but it would make it easier for the users of the repo to use the repo. True it is not unreasonable to make them install the other repos but it would also be nice to have the workflow pass here.
> 
> 
> 
> I am not sure if you had to make changes to the make depend script or not. It seems not and I think having in the make dependencies the full path of include files is not ideal but of course the above idea would also make this not be a problem (at least as I think in my tired head).

Please remember that `dbg` and `dyn_array` are general libraries that are used in a wide variety of applications.  We don't want to start "ifdef-ing" every use in the set.

It was bad enough that we had to use `MKIOCCCENTRY_SRC`.  Doing that you proposed would make things worse from the special case code.

We probably should have renamed `MKIOCCCENTRY_SRC` to something like `LOCAL_INCLUDE` .. we might do that just to make the change more generic.

**Comment by xexyl** [comment]
> > It might be nice if we could make use of the system used in mkiocccentry so that the workflows can pass and so that people can more easily compile this repo. For instance it might be something like (I am typing this with bleary eyes btw):
> > ```c
> > #ifdef JPARSE_REPO
> > 
> > #include "dbg/dbg.h"
> > 
> > /* etc. */
> > 
> > #elif defined (MKIOCCCENTRY_SRC)
> > 
> > #include "../dbg/dbg.h"
> > 
> > /* etc. */
> > 
> > #endif 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This would not only make it easy for the workflow to work right but it would make it easier for the users of the repo to use the repo. True it is not unreasonable to make them install the other repos but it would also be nice to have the workflow pass here.
> > I am not sure if you had to make changes to the make depend script or not. It seems not and I think having in the make dependencies the full path of include files is not ideal but of course the above idea would also make this not be a problem (at least as I think in my tired head).
> 
> Please remember that `dbg` and `dyn_array` are general libraries that are used in a wide variety of applications. We don't want to start "ifdef-ing" every use in the set.
> 

That's true. See below.

> It was bad enough that we had to use `MKIOCCCENTRY_SRC`. Doing that you proposed would make things worse from the special case code.

That is true but without doing this the workflows will fail at least until the make dependencies have system paths.  If you have a solution to that then please feel free to suggest one.

**Comment by xexyl** [comment]
> > That is true but without doing this the workflows will fail at least until the make dependencies do not have system paths. If you have a solution to that then please feel free to suggest one.
> 
> The solution in the workflow is to clone and install the repositories. See [test.yml from mkiocccentry](https://github.com/ioccc-src/mkiocccentry/blob/master/.github/workflows/test.yml).
> 
> Do stuff such as:
> 
> ```yaml
>     - name: clone dbg
>       run: git clone https://github.com/lcn2/dbg
>     - name: install dbg
>       run: cd dbg && sudo make clobber all install
>     - name: clone dyn_array
>       run: git clone https://github.com/lcn2/dyn_array
>     - name: install dyn_array
>       run: cd dyn_array && sudo make clobber all install
> ```
> 
> This will install stuff in the test system for jparse and friends to find.

Pretty sure I did that yesterday ... no luck if I did.

**Comment by xexyl** [comment]
Oh .. I know. I made a yank paste error. I might have a fix.

**Comment by xexyl** [comment]
Well it still runs into a problem:

```
Run make
cc -std=gnu17 -O3 -g3 -pedantic -Wall -Wextra -Wno-unused-but-set-variable -Wno-char-subscripts   verge.c -c
cc -std=gnu17 -O3 -g3 -pedantic -Wall -Wextra -Wno-unused-but-set-variable -Wno-char-subscripts   util.c -c
make: *** No rule to make target '../dbg/Makefile', needed by '../dbg/dbg.a'.  Stop.
Error: Process completed with exit code 2.
```

Maybe I need to look at this again when I'm not as tired.

**Comment by xexyl** [comment]
... the dynamic array did seem to work though whether the next steps would have succeeded or not I can't say since the dbg step failed. But at least that did install.

**Comment by xexyl** [comment]
Ah! It is not the dbg repo not being installed; it is that the make of this repo fails due to the dbg repo Makefile. But why?

**Comment by xexyl** [comment]
I can simulate this. By not having a directory `../dbg/` it causes the problem.

**Comment by xexyl** [comment]
And the answer seems to be because the Makefile has things like:

```makefile
jparse: jparse_main.o jparse.a ../dyn_array/dyn_array.a ../dbg/dbg.a
```

It would seem good to update it if a change has happened but the way this works right now means that one requires a `../dbg` and `../dyn_array`. This is obviously a problem.

**Comment by xexyl** [comment]
> And the answer seems to be because the Makefile has things like:
> 
> ```makefile
> jparse: jparse_main.o jparse.a ../dyn_array/dyn_array.a ../dbg/dbg.a
> ```
> 
> It would seem good to update it if a change has happened but the way this works right now means that one requires a `../dbg` and `../dyn_array`. This is obviously a problem.

In fact there are a lot of places where these external libraries are referred to. Something has to be done here and I don't know what it is. Maybe after your sleep you have some ideas. It would, as noted, simplify it for users if the repo had a subdirectory for each, but of course this makes it messier esp if mkiocccentry also (as it does) require these repos.

**Comment by xexyl** [comment]
It would also be good if we could not have system paths included in the make dependencies. But that is a thing that is from your tool that builds them. 

I don't know. This seems problematic at this time. We can discuss these details later on. I'll be away for a while now too. Unfortunately no chance or very little chance of more sleep and I woke up earlier today but hopefully things will go well.

**Comment by xexyl** [comment]
I'm starting to think it mighta actually be a macOS thing and in particular it must not consider `/usr/local/include` system path? Or maybe it is that in all platforms even? Since looking more carefully I see that the make depend rule does use `cc -MM` .... So this part might not be possible to fix unless we filter out output that has `/usr/local` as well (starting with it).

**Comment by xexyl** [comment]
In linux this happens without the library installed:

```
make: *** No rule to make target '/usr/local/include/dbg.h', needed by 'verge.o'.  Stop.
```

Of course the Makefile has for that target:

```makefile
verge.o: verge.c verge.h
        ${CC} ${CFLAGS} verge.c -c

verge: verge.o util.o ../dbg/dbg.a ../dyn_array/dyn_array.a
        ${CC} ${CFLAGS} $^ -o $@
```

So that seems to be a problem too, possibly.

You said before maybe we need `-ldbg` and the like. Perhaps it's time to do that? Though still not sure about the header files being dependencies or not.

**Comment by xexyl** [comment]
Yes this system paths thing is a macOS problem. After installing the libraries and the independ tool in linux I ran `make depend` and now I see nothing of the sort including `/usr/local/include` :((

Instead the Makefile now shows for example:

```makefile
jparse.o: jparse.c jparse.h jparse.tab.h json_parse.h json_sem.h \
    json_util.h util.h
```

**Comment by xexyl** [comment]
In order to do proper linking we have to rename the other libraries to `libdbg.a` and `libdyn_array.a` so that `-ldbg` and `-ldyn_array` will work. In linux one can do say: `-l:dbg.a` but that is not supported in macOS.

Now I can do this if you want (though later .. going afk shortly) but this is another thing that we have to sort out I think since right now the build here assumes that the dbg and dyn_array directories exist.

**Comment by xexyl** [comment]
Meanwhile this diff has to be applied (or something else has to change?):

```diff
@@ -502,16 +502,10 @@ run_flex-v7: verge sorry.tm.ca.h
 
 #########################################################
 # rules that invoke Makefile rules in other directories #
 #########################################################
 
-../dbg/dbg.a: ../dbg/Makefile
-       ${Q} ${MAKE} ${MAKE_CD_Q} -C ../dbg extern_liba C_SPECIAL=${C_SPECIAL}
-
-../dyn_array/dyn_array.a: ../dyn_array/Makefile
-       ${Q} ${MAKE} ${MAKE_CD_Q} -C ../dyn_array extern_liba C_SPECIAL=${C_SPECIAL}
-
```

Afk now ... will look on this later on. Have other things I have to do now.

**Comment by xexyl** [comment]
With https://github.com/lcn2/dyn_array/pull/3/commits/c8144fb58bc5675ca1dd558d3537473cfb0deb3f in the dyn_array repo and with commits https://github.com/lcn2/dbg/pull/21/commits/2d9a3a537bafdeaffc47d53d258381f07e2d349d and https://github.com/lcn2/dbg/pull/21/commits/832f7e46b6f44e08f7bd4417882de23fa69333cf as well as the commit https://github.com/ioccc-src/mkiocccentry/pull/949/commits/2a5f5023e38f114c80de39f1c3477d912598b991 in the mkiocccentry repo the libraries have been renamed to `libdyn_array.a` and `libdbg.a` so that one can link them without having to have the library files in the same directory, important for this repo.

The mkiocccentry just has it updated so that it matches the other two except that it does **NOT** use `-L. -ldbg` or `-L. -ldyn_array`. If you wish for it to do that then I can do that too.

Once the commits to the two libraries have been merged I can make some changes to the jparse Makefiles and hopefully this will allow one to compile this repo on its own. I might make the changes to the Makefile now but if I do that it won't pass and it might be better to wait so we can verify that all is okay. On the other hand I can update the README.md at that point and that would be sufficient.

I hope this helps.

**Comment by xexyl** [comment]
With commit 3975b25811cf73bc64907fccf6cb0a1531f9ff10 the Makefiles have been PREPARED for the renaming of the library files. Once those are done it should, I believe, be possible to install and use this library without having `../dbg/` and `../dyn_array/` directories with the libraries built.

At that time I will update the README.md with details on how to get, build and compile the required libraries. This will of course end up testing the workflow.

If all is okay then I believe this issue can be marked resolved. However, I will leave that to your decision as perhaps you have some ideas for improvement.

**Comment by xexyl** [comment]
I will be afk a couple hours or so. When back I have other things to take care of but hopefully we can finish this task today.

**Comment by xexyl** [comment]
An important fix in the mkiocccentry commit was changed. It required a new commit entirely. See https://github.com/ioccc-src/mkiocccentry/pull/950/commits/bd15af4f2d2887948d3f9bde8819701dad02e168.

Fortunately I had a backup of the git repo itself! If not I would have possibly run into problems as I had to use that copy and then do a pull and then copy the modified files back into it.

**Comment by xexyl** [comment]
Meanwhile I updated the README.md here for when the other repos have been updated. I also updated the lib filename in the Makefile but these I will not commit until the other repos are okay so that I can test things out.

I finished the year 2005 in the website also and that's all I'll do today. Made a lot of progress and I am thrilled! Tomorrow we can discuss further details and make any corrections if necessary.

**Comment by lcn2** [comment]
> Oh .. I know. I made a yank paste error. I might have a fix.

The YML configuration is very unforgiving, but once you lock in something that works .. and until some radically changes, you will be "good to go" as the expression goes.

**Comment by lcn2** [comment]
> It would also be good if we could not have system paths included in the make dependencies. But that is a thing that is from your tool that builds them.
> 
> I don't know. This seems problematic at this time. We can discuss these details later on. I'll be away for a while now too. Unfortunately no chance or very little chance of more sleep and I woke up earlier today but hopefully things will go well.

You raise a good point.  Perhaps the a grep if needed to exclude any "/usr/include/" and any "/usr/local/include" paths from going into `make depend` output.

**Comment by lcn2** [comment]
We will need time get to the updates you have made and perhaps we have proposed after we incorporate whatever PRs you have issued.

A key thing will be to stay on top of changes to the dbg, dyn_array and jparse repo changes .. especially when it comers to how they import into mkiocccentry.   Nevertheless, once these repos settle down, and things become coherent, we will be good to go.  We just need to get over a few "bumps" along the way.

**Comment by lcn2** [comment]
You will need to update the dbg and dyn_alloc repos and perform the `make clobber all test install` actions.

With commit 726c13cded90a51fc143332059c5aaee78eb7c68 in the mkiocccentry repo and PR #6 in this repo, we believe that the issues raised in this issue #2 have been resolved.

With the above actions, we have tested the result under both macOS and RHEL9 Linux.  As such we believe you make close this issue #2.

**Comment by xexyl** [comment]
I am not sure if it's done just yet because the workflow failed. But I might know why. If so it should be an easy fix. I will have laptop on soon and then I can look into that. 

Thanks!

**Comment by xexyl** [comment]
Unfortunately it seems like not everything was done in dyn_array ... so have to wait for the fix to be merged. I just made some additional fixes here. When that other repo is fixed I can push a test here. In the meantime I will keep this open.

But we're about there I think! Thanks for the help so far.

**Comment by xexyl** [comment]
So annoying I missed those places .. I want this done. I'll be afk a while but when back I have other things to do .. then maybe can look at the next year on the website.

**Comment by xexyl** [comment]
I fixed codeql.yml to install the dependencies but the problem is I forgot a reference in dyn_array Makefile to the old dbg filename so it fails to link. Thus I am waiting on the fix I made earlier today to be merged. It is hoped at this point the tests will pass.

I have updated the README.md file to explain how to install the other repos and then build this repo but I am waiting for the other repos to work okay. I can then close this. Hopefully you can get to it earlier today but if not I'll push the changes tomorrow.

No problem if you cannot of course but it'd be nice if we can get this resolved.

**Comment by xexyl** [comment]
Of course we could close this issue as resolved as it certainly should be but it would be nice if we could have GitHub do it too. I have other things to take care of and I hope to look at 2006 'over there' after that. Thanks for all the work here! It's appreciated **VERY MUCH**!

**Comment by xexyl** [comment]
Hmm ... shouldn't all the header files be installed with `make install`? And if so ought we to have a subdirectory `jparse` where they reside? Or if not they might be prefixed (in name) with `jparse_`? A subdirectory seems cleaner to debut I can imagine someone wanting to use the library will want to use the functions in say `util.c`. Of course they exist in the library but they will be implicit function declarations, unless I'm not thinking of something.

That would be another reason to keep this open though maybe it should be a new issue of make it easier to use.

**Comment by lcn2** [comment]
> Hmm ... shouldn't all the header files be installed with `make install`? And if so ought we to have a subdirectory `jparse` where they reside? Or if not they might be prefixed (in name) with `jparse_`?

Yes, they probably should.  One day to preserve the include filenames is to install them under a `jparse/` sub directory as in `/usr/local/include/jparse/something.h`.   While you have to change include locations the sub-directory approach is much cleaner.

**Comment by xexyl** [comment]
The workflow has passed!

**Comment by xexyl** [comment]
> > Hmm ... shouldn't all the header files be installed with `make install`? And if so ought we to have a subdirectory `jparse` where they reside? Or if not they might be prefixed (in name) with `jparse_`?
> 
> Yes, they probably should. One day to preserve the include filenames is to install them under a `jparse/` sub directory as in `/usr/local/include/jparse/something.h`. While you have to change include locations the sub-directory approach is much cleaner.

I think subdirectory is much better too.

> 
> ## UPDATE 0
> If you rename files or better yet move installed files, please keep in home the "self-contained" nature of the mkiocccentry repo, perhaps missing a PR if needed to adjust those Makefiles to still use the `INTERNAL_INCLUDE` method.

Please elaborate this part. Or maybe you'd like to take care of this? I'd be happy to update the man pages and README.md file.

**Comment by xexyl** [comment]
I believe the `make install` could just install them to the `/usr/local/include/jparse` directory and then we could in the Makefiles use `-I` to address the local build .. though I haven't given that much thought. What do you think?

**Comment by lcn2** [comment]
> I believe the `make install` could just install them to the `/usr/local/include/jparse` directory and then we could in the Makefiles use `-I` to address the local build .. though I haven't given that much thought. What do you think?

Yes.

**Comment by lcn2** [comment]
> > ## UPDATE 0
> > If you rename files or better yet move installed files, please keep in home the "self-contained" nature of the mkiocccentry repo, perhaps missing a PR if needed to adjust those Makefiles to still use the `INTERNAL_INCLUDE` method.
> 
> Please elaborate this part. Or maybe you'd like to take care of this? I'd be happy to update the man pages and README.md file.

We are saving that you should check on the effect of such changes on the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry), and if needed issue a PR like we did for

- commit 6ef2b7781648d9636d7253c4fa07bb6284663c8e

where we did:

```sh
git status # to be sure the tree is clean
make release # to see if things can still build and test
make clobber # to be sure the tree is clean

make dyn_array.help
make dyn_array.recreate_clone
make dyn_array.update_from_clone

git status # to see if anything changed
make release # to see if things can still build and test
```

We then did an update for the repo because things changed and things can still build.

BTW, we also did:

```sh
git status # to be sure the tree is clean
make release # to see if things can still build and test
make clobber # to be sure the tree is clean

make jparse.help
make jparse.recreate_clone
make jparse.update_from_clone

git status # to see if anything changed
make release # to see if things can still build and test
```

Because nothing changed, now update to the repo was needed.

**Comment by xexyl** [comment]
> > > ## UPDATE 0
> > > If you rename files or better yet move installed files, please keep in home the "self-contained" nature of the mkiocccentry repo, perhaps missing a PR if needed to adjust those Makefiles to still use the `INTERNAL_INCLUDE` method.
> > 
> > 
> > Please elaborate this part. Or maybe you'd like to take care of this? I'd be happy to update the man pages and README.md file.
> 
> We are saving that you should check on the effect of such changes on the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry), and if needed issue a PR like we did for
> 
> * commit 6ef2b7781648d9636d7253c4fa07bb6284663c8e
> 
> where we did:
> 
> ```shell
> git status # to be sure the tree is clean
> make release # to see if things can still build and test
> make clobber # to be sure the tree is clean
> 
> make dyn_array.help
> make dyn_array.recreate_clone
> make dyn_array.update_from_clone
> 
> git status # to see if anything changed
> make release # to see if things can still build and test
> ```
> 
> We then did an update for the repo because things changed and things can still build.
> 
> BTW, we also did:
> 
> ```shell
> git status # to be sure the tree is clean
> make release # to see if things can still build and test
> make clobber # to be sure the tree is clean
> 
> make jparse.help
> make jparse.recreate_clone
> make jparse.update_from_clone
> 
> git status # to see if anything changed
> make release # to see if things can still build and test
> ```
> 
> Because nothing changed, now update to the repo was needed.
> 
> ## UPDATE 0
> We are suggesting you can do the above to look at the impact of a jparse change on the mkiocccentry repo, and adjust and/or issue PRs as needed.

Ah I see.

My only concern with this suggested change is that I'm not sure what path(s) for `-I`. My biggest fear is that changing the path by installing to `/usr/local/jparse` I might break local directory build as I would have the system installed files already there .. or maybe I wouldn't. I guess I can sort this out. Doing something else first but then can hopefully look at it.

Maybe I should quickly open an issue? That way this one can be closed as complete and we can discuss any necessary changes.

**Comment by xexyl** [comment]
GitHub yaml parsing is so picky ...  In https://github.com/xexyl/jparse/actions/runs/10638041246/workflow:

```
You have an error in your yaml syntax on line 37
```

Line 37 is:

```yaml
  analyze:
```

At first I had it as 'analyse' but I seemed to recall it complaining about British English so I changed it. That did not solve the problem (which makes sense). What does not make sense is that this detects a syntax error.

Any idea?

**Comment by xexyl** [comment]
> > > We are suggesting you can do the above to look at the impact of a jparse change on the mkiocccentry repo, and adjust and/or issue PRs as needed.
> > 
> > 
> > Ah I see.
> > My only concern with this suggested change is that I'm not sure what path(s) for `-I`. My biggest fear is that changing the path by installing to `/usr/local/jparse` I might break local directory build as I would have the system installed files already there .. or maybe I wouldn't. I guess I can sort this out. Doing something else first but then can hopefully look at it.
> > Maybe I should quickly open an issue? That way this one can be closed as complete and we can discuss any necessary changes.
> 
> So you should perform the actions as described on [GH-issuecomment-2322179729](https://github.com/xexyl/jparse/issues/2#issuecomment-2322179729) and see what happens to the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry), adjusting or fixing or improving if/as needed including possible PRs. The procedure to bring the code from other repos and check has been make very easy and quick. We recommend you use it when considering such a change that might or might not break a further [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) updates.

I will have to do that yes. But of course first I need to fix this bizarre syntax error. It seems so strange. Funny how nothing shown in syntax highlighting (as an error which I think GitHub shows) and in vim no problems either but in workflows it apparently is a syntax error. I wonder if it's one of those cases where a surrounding line is the problem.

**Comment by xexyl** [comment]
What's amusing is that the GitHub workflow editor also shows no syntax errors.

**Comment by lcn2** [comment]
> GitHub yaml parsing is so picky ... In https://github.com/xexyl/jparse/actions/runs/10638041246/workflow:
> 
> ```
> You have an error in your yaml syntax on line 37
> ```
> 
> Line 37 is:
> 
> ```yaml
>   analyze:
> ```
> 
> At first I had it as 'analyse' but I seemed to recall it complaining about British English so I changed it. That did not solve the problem (which makes sense). What does not make sense is that this detects a syntax error.
> 
> Any idea?

A diff with the [similar YML file](https://github.com/ioccc-src/mkiocccentry/blob/master/.github/workflows/codeql.yml) from the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) shows:

```diff
37,38c37,38
<   analyze:
<     name: Analyze
---
>   analyse:
>     name: Analyse
```

hmm ...

**Comment by xexyl** [comment]
> > GitHub yaml parsing is so picky ... In https://github.com/xexyl/jparse/actions/runs/10638041246/workflow:
> > ```
> > You have an error in your yaml syntax on line 37
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Line 37 is:
> > ```yaml
> >   analyze:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > At first I had it as 'analyse' but I seemed to recall it complaining about British English so I changed it. That did not solve the problem (which makes sense). What does not make sense is that this detects a syntax error.
> > Any idea?
> 
> A diff with the [similar YML file](https://github.com/ioccc-src/mkiocccentry/blob/master/.github/workflows/codeql.yml) from the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) shows:
> 
> ```diff
> 37,38c37,38
> <   analyze:
> <     name: Analyze
> ---
> >   analyse:
> >     name: Analyse
> ```
> 
> hmm ...

Yes and I just changed it to 'Anaylze'.

**Comment by lcn2** [comment]
> What's amusing is that the GitHub workflow editor also shows no syntax errors.

Under vim, the jparse `analyse:` is colored green but under mkiocccentry it is coloured blue.

**Comment by xexyl** [comment]
I know spaces caused a problem before but here's the thing: vimdiff shows NO difference!

**Comment by xexyl** [comment]
> > What's amusing is that the GitHub workflow editor also shows no syntax errors.
> 
> Under vim, the jparse `analyse:` is colored green but under mkiocccentry it is coloured blue.

That's because you have American English as your language.

**Comment by xexyl** [comment]
Maybe I need to recopy it over and just make the changes. But I basically did that except removing the part that it said to remove.

**Comment by lcn2** [comment]
> > > What's amusing is that the GitHub workflow editor also shows no syntax errors.
> > 
> > 
> > Under vim, the jparse `analyse:` is colored green but under mkiocccentry it is coloured blue.
> 
> That's because you have American English as your language.

No to mention our spelling pub we put in our reply.  😉

**Comment by xexyl** [comment]
> > > > What's amusing is that the GitHub workflow editor also shows no syntax errors.
> > > 
> > > 
> > > Under vim, the jparse `analyse:` is colored green but under mkiocccentry it is coloured blue.
> > 
> > 
> > That's because you have American English as your language.
> 
> Not to mention our spelling pub we put in our reply. 😉

But that's not British English specific. If you mean the word .. ah you just fixed it.

**Comment by xexyl** [comment]
> > > > What's amusing is that the GitHub workflow editor also shows no syntax errors.
> > > 
> > > 
> > > Under vim, the jparse `analyse:` is colored green but under mkiocccentry it is coloured blue.
> > 
> > 
> > That's because you have American English as your language.
> 
> Not to mention our spelling pun we put in our reply. 😉

Or at least double spelling which is in a sense wordplay: at least it's ironic since you're talking about different spellings in vim.

**Comment by lcn2** [comment]
> > > > > What's amusing is that the GitHub workflow editor also shows no syntax errors.
> > > > 
> > > > 
> > > > Under vim, the jparse `analyse:` is colored green but under mkiocccentry it is coloured blue.
> > > 
> > > 
> > > That's because you have American English as your language.
> > 
> > 
> > Not to mention our spelling pub we put in our reply. 😉
> 
> But that's not British English specific. If you mean the word .. ah you just fixed it.

We decided that putting meta-puns in our note about pubs was too meta. 🤓

**Comment by xexyl** [comment]
> > > > > > What's amusing is that the GitHub workflow editor also shows no syntax errors.
> > > > > 
> > > > > 
> > > > > Under vim, the jparse `analyse:` is colored green but under mkiocccentry it is coloured blue.
> > > > 
> > > > 
> > > > That's because you have American English as your language.
> > > 
> > > 
> > > Not to mention our spelling pub we put in our reply. 😉
> > 
> > 
> > But that's not British English specific. If you mean the word .. ah you just fixed it.
> 
> We decided that putting meta-puns in our note about pubs was too meta. 🤓

But hopefully not Meta!

**Comment by lcn2** [comment]
> > > GitHub yaml parsing is so picky ... In https://github.com/xexyl/jparse/actions/runs/10638041246/workflow:
> > > ```
> > > You have an error in your yaml syntax on line 37
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Line 37 is:
> > > ```yaml
> > >   analyze:
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > At first I had it as 'analyse' but I seemed to recall it complaining about British English so I changed it. That did not solve the problem (which makes sense). What does not make sense is that this detects a syntax error.
> > > Any idea?
> > 
> > 
> > A diff with the [similar YML file](https://github.com/ioccc-src/mkiocccentry/blob/master/.github/workflows/codeql.yml) from the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) shows:
> > ```diff
> > 37,38c37,38
> > <   analyze:
> > <     name: Analyze
> > ---
> > >   analyse:
> > >     name: Analyse
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > hmm ...
> 
> Yes and I just changed it to 'Anaylze'.

Feel free to file a bug report with GitHub .. it should be easy for them to allow both spellings.

**Comment by xexyl** [comment]
> > > > GitHub yaml parsing is so picky ... In https://github.com/xexyl/jparse/actions/runs/10638041246/workflow:
> > > > ```
> > > > You have an error in your yaml syntax on line 37
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > Line 37 is:
> > > > ```yaml
> > > >   analyze:
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > At first I had it as 'analyse' but I seemed to recall it complaining about British English so I changed it. That did not solve the problem (which makes sense). What does not make sense is that this detects a syntax error.
> > > > Any idea?
> > > 
> > > 
> > > A diff with the [similar YML file](https://github.com/ioccc-src/mkiocccentry/blob/master/.github/workflows/codeql.yml) from the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) shows:
> > > ```diff
> > > 37,38c37,38
> > > <   analyze:
> > > <     name: Analyze
> > > ---
> > > >   analyse:
> > > >     name: Analyse
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > hmm ...
> > 
> > 
> > Yes and I just changed it to 'Anaylze'.
> 
> Feel free to file a bug report with GitHub .. it should be easy for them to allow both spellings.

I'm not sure if that is it though. Oddly I just tried to resync from the mkiocccentry repo, just updating the necessary part, and I don't see the CodeQL at all in a new run. Yet previously when I committed the file it did run it.

**Comment by xexyl** [comment]
Just started a manual run.

**Comment by xexyl** [comment]
Okay it worked. That is one of the strangest things ever ....

**Comment by lcn2** [comment]
When this issue is closed (and any changes to [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) that might be needed are applied), it might be time to update [mkiocccentry CHANGES.md](https://github.com/ioccc-src/mkiocccentry/blob/master/CHANGES.md) (including a revised `MKIOCCCENTRY_REPO_VERSION` value).

If that is done and things go quiet for a few days, a new [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) would be in order.

**Comment by xexyl** [comment]
> When this issue is closed (and any changes to [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) that might be needed are applied), it might be time to update [mkiocccentry CHANGES.md](https://github.com/ioccc-src/mkiocccentry/blob/master/CHANGES.md) (including a revised `MKIOCCCENTRY_REPO_VERSION` value).
> 
> If that is done and things go quiet for a few days, a new [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) would be in order.

That sounds like a great idea.

**Comment by xexyl** [comment]
To see how bizarre the fix to the CodeQL file is:

```diff
diff --git a/.github/workflows/codeql.yml b/.github/workflows/codeql.yml
index 0b09fdae36df79fbd2832323486d64bc40be109a..5832bbfc4ef9d44aa8b2955432b0ac5afdcf2107 100644
--- a/.github/workflows/codeql.yml
+++ b/.github/workflows/codeql.yml
@@ -44,21 +44,33 @@ jobs:
 
     strategy:
       fail-fast: false
       matrix:
         language: [ 'cpp' ]
+        # CodeQL supports [ 'cpp', 'csharp', 'go', 'java', 'javascript', 'python', 'ruby' ]
+        # Learn more about CodeQL language support at https://aka.ms/codeql-docs/language-support
 
     steps:
     - name: Checkout repository
       uses: actions/checkout@v4
 
+    # Initializes the CodeQL tools for scanning.
     - name: Initialize CodeQL
       uses: github/codeql-action/init@v3
       with:
         languages: ${{ matrix.language }}
        
-     - run: |
+        # Details on CodeQL's query packs refer to : https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-code-scanning#using-queries-in-ql-packs
+        # queries: security-extended,security-and-quality
+
+    # ℹ️ Command-line programs to run using the OS shell.
+    # 📚 See https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsrun
+
+    #   If the Autobuild fails above, remove it and uncomment the following three lines.
+    #   modify them (or add more) to build your code if your project, please refer to the EXAMPLE below for guidance.
+
+    - run: |
        git clone https://github.com/lcn2/dbg
        cd dbg
        make clobber all
        sudo make install
        git clone https://github.com/lcn2/dyn_array

```

.. which is one of the funniest things I have seen in a while.

**Comment by xexyl** [comment]
Hmm ... to save the trouble (I'm tired and I need to do some other things) maybe you can answer this question that should be answerable to me immediately but is not right now.

If I just make the make install add the relevant header files (not all as some are for tools) to `/usr/local/include/jparse` then I can update the documentation to have the `jparse/` as part of the include paths (for those using the API) and that's all I should have to do. Is that right or am I forgetting something?

The only downside to this of course is that even in the include directory we will have `"` instead of `<>`. Maybe I should update the Makefile to have `-I.` and then update the include files to use `<>`?

What do you think on both parts?

**Comment by xexyl** [comment]
Added todo at top comment for the comment you included what you need done for the mkiocccentry repo. Until the two things in the above comment (that I just made) are answered/resolved I will not do that task. Once they are I can do the final task and if all is good close this issue.

**Comment by xexyl** [comment]
> To see how bizarre the fix to the CodeQL file is:
> 
> ```diff
> diff --git a/.github/workflows/codeql.yml b/.github/workflows/codeql.yml
> index 0b09fdae36df79fbd2832323486d64bc40be109a..5832bbfc4ef9d44aa8b2955432b0ac5afdcf2107 100644
> --- a/.github/workflows/codeql.yml
> +++ b/.github/workflows/codeql.yml
> @@ -44,21 +44,33 @@ jobs:
>  
>      strategy:
>        fail-fast: false
>        matrix:
>          language: [ 'cpp' ]
> +        # CodeQL supports [ 'cpp', 'csharp', 'go', 'java', 'javascript', 'python', 'ruby' ]
> +        # Learn more about CodeQL language support at https://aka.ms/codeql-docs/language-support
>  
>      steps:
>      - name: Checkout repository
>        uses: actions/checkout@v4
>  
> +    # Initializes the CodeQL tools for scanning.
>      - name: Initialize CodeQL
>        uses: github/codeql-action/init@v3
>        with:
>          languages: ${{ matrix.language }}
>         
> -     - run: |
> +        # Details on CodeQL's query packs refer to : https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-code-scanning#using-queries-in-ql-packs
> +        # queries: security-extended,security-and-quality
> +
> +    # ℹ️ Command-line programs to run using the OS shell.
> +    # 📚 See https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsrun
> +
> +    #   If the Autobuild fails above, remove it and uncomment the following three lines.
> +    #   modify them (or add more) to build your code if your project, please refer to the EXAMPLE below for guidance.
> +
> +    - run: |
>         git clone https://github.com/lcn2/dbg
>         cd dbg
>         make clobber all
>         sudo make install
>         git clone https://github.com/lcn2/dyn_array
> ```
> 
> .. which is one of the funniest things I have seen in a while.

Actually .. aren't there stories of code that segfaulted without certain comments?

**Comment by xexyl** [comment]
I will be leaving for the day but tomorrow morning I will take care of the above pending issues (with header files) and then the mkiocccentry repo, assuming that the questions are all resolved. Hopefully at that point we can consider this repo in a good state and won't need to work on it until the new tools are to be written. At that point I can get back to the website.

Anyway good day!

**Comment by lcn2** [comment]
> The only downside to this of course is that even in the include directory we will have " instead of <>. Maybe I should update the Makefile to have -I. and then update the include files to use <>?
> 
> What do you think on both parts?

Give the above a try and see how it goes. 

The dbg and dyn_array repos are intended to be standalone, install under `/usr/local/include` as used by other code that is well outside of these repos.  The use of <> by default is proper here. 

> Yes, we might have been better off using `/usr/local/include/dbg` and `/usr/local/include/dyn_array/` but those "Oliphaunts sailed out of barn door and thru the docks" long ago, as the mixed expression goes, so we can't change that. 

The Makefiles in mkiocccetry could use the `-I` trick as a result.  Give it a try.

**Comment by xexyl** [comment]
Well I can do that, the `<>` but I'm not sure how well it'll work without `-I.`. That might be better. The man pages use that, at least for `#include <jparse/jparse.h>` but of course this repo cannot have the `jparse/` in front of the files as the header files are only in a subdirectory when installed, not otherwise.

As for the Oliphaunts comment: thanks! That was lovely.

Going back to the `<>` I might play with that but not today as I have other things I need to do. Going to do a commit in the mkiocccentry repo, updating the release, and then be done with that for the day. As it is I'm not sure if I can finish the other things I need to do today before I go (but fortunately they can be delayed).

Perhaps tomorrow I can play with the `<>` instead of `"` in the files here. In the meantime I think I will be able to mark this issue as complete! I will have you give it a go though. I did a test outside this repo for system installed header files too, btw, and all is good!

The uninstall rule has a joke that you might appreciate .. a message Easter egg you might say :-)

**Comment by xexyl** [comment]
If https://github.com/ioccc-src/mkiocccentry/pull/952/commits/574ff8a2cf255601e0e9392dc80652082e49baf5 seems good to you feel free to close this issue as complete.

Unless you wish me to look at the `<>` of course .. or you come across something else. 

.. also just made a fix to that commit .. hopefully passes workflow now. An oversight on my part.

**Comment by xexyl** [comment]
Okay the `-L` paths are fixed now so that the workflow passes in that repo (and anyone who does not have the dyn_array and dbg libraries installed also would be affected). 

If everything looks good please feel free to either close this as complete or let me know and I can do it. Off to do other things.

**Comment by xexyl** [comment]
Oh quickly as far as `<>` mode.

I believe (but you certainly know better than me) that:

0. You can always use `""` even for standard locations.
1. `""` works in local directories whereas `<>` does not.
2. If you use `-I` it allows for `<>` even in that directory if it's not a system directory.

Is that correct?

**Comment by lcn2** [comment]
> Oh quickly as far as `<>` mode.
> 
> I believe (but you certainly know better than me) that:
> 
> 0. You can always use `""` even for standard locations.
> 1. `""` works in local directories whereas `<>` does not.
> 2. If you use `-I` it allows for `<>` even in that directory if it's not a system directory.
> 
> Is that correct?

Yes:

```
       -I<directory>
              Add the specified directory to the search path for include files.
```

**Comment by lcn2** [comment]
We believe with the recent changes that this issue has been resolved.  Thank you @xexyl

**Comment by xexyl** [comment]
> > Oh quickly as far as `<>` mode.
> > I believe (but you certainly know better than me) that:
> > 
> > 0. You can always use `""` even for standard locations.
> > 1. `""` works in local directories whereas `<>` does not.
> > 2. If you use `-I` it allows for `<>` even in that directory if it's not a system directory.
> > 
> > Is that correct?
> 
> Yes:
> 
> ```
>        -I<directory>
>               Add the specified directory to the search path for include files.
> ```

Right. I've seen that many times of course. But the other parts I either read or discovered. Glad this issue is resolved. Thanks for the opening and the suggestions and help!

Back to what I was doing .. tomorrow I can do more.

**Comment by xexyl** [comment]
Why did you open it? Please elaborate and I will look at it tomorrow. 

Thanks.

**Comment by lcn2** [comment]
We suggest that references to `-L../dbg -L../dyn_array` be addressed in the `Makefile` as they require the dbg and dyn_array libraries to be located in a parallel directory.

**Comment by xexyl** [comment]
Well the `-L` is just additional directories, right?

I will have to read more tomorrow. Sorry for the delays!

**Comment by lcn2** [comment]
> Well the `-L` is just additional directories, right?
> 
> I will have to read more tomorrow. Sorry for the delays!

We save a fix to the problems noted in [GH-issuecomment-2323043643](https://github.com/xexyl/jparse/issues/2#issuecomment-2323043643) that we are testing right now.  Look for a PR very soon.

**Comment by lcn2** [comment]
We have a commit 65e8a9d41584771b679ad4e5dbef18be3ef04090 in our [jparse forked repo](https://github.com/lcn2/jparse/tree/master) that we plan to test out RHEL 9 Linux and then on the mkiocccentruy repo.   Stay tuned ...

**Comment by lcn2** [comment]
We recommend that you update the [dbg repo](https://github.com/lcn2/dbg).   An install would be a good thing to do.

We recommend that you update the [dyn_arrau repo](https://github.com/lcn2/dyn_array).    An install would be a good thing to do.

Please apply PR #8.

**Comment by xexyl** [comment]
Merged it a bit ago. I will make the other changes tomorrow morning. Well `make` them.

**Comment by xexyl** [comment]
If you want to close this issue as resolved you can. Otherwise we can wait until my updates tomorrow. 

Hope you have a great night! Must sleep now.

**Comment by xexyl** [comment]
I will be busy for a bit and then I will work on the PREFIX install related things. 

I do need to find the energy to do some house cleaning but with the temperature I might not manage to do much of that. I helped myself out by doing some last night but some mopping needs to be done too. 

If I don't get to that today it's not the end of the world (those pesky Vogons 🛸👽☄️🪐🚀🧑‍🚀though .. there's no accounting for them .. so maybe I don't have to worry about them: I mean Arthur Dent didn't have to so maybe I don't either!)

Well anyway I will get this stuff here done today if nothing else. 

But afk a bit.

**Comment by lcn2** [comment]
> If I don't get to that today it's not the end of the world (those pesky Vogons 🛸👽☄️🪐🚀🧑‍🚀though .. there's no accounting for them .. so maybe I don't have to worry about them: I mean Arthur Dent didn't have to worry about his house so maybe we don't either!)
> 
> Well anyway I will get this stuff here done today if nothing else. But afk a bit.

Just as long as you avoid poetry from Grunthos tthe Flatulent today, you should do fine! 😁

**Comment by xexyl** [comment]
> > If I don't get to that today it's not the end of the world (those pesky Vogons 🛸👽☄️🪐🚀🧑‍🚀though .. there's no accounting for them .. so maybe I don't have to worry about them: I mean Arthur Dent didn't have to worry about his house so maybe we don't either!)
> > Well anyway I will get this stuff here done today if nothing else. But afk a bit.
> 
> Just as long as you avoid poetry from Grunthos the Flatulent 💨 today, you should do fine! 😁

I suppose so .. but I had forgotten about him. Still: thanks for the laugh! I did not, in any case, finish what I had hoped to, and today is going to be hectic. Worker will be here and I have a zoom call. Then I have the other things to do too. I hope I can finish those things but not sure if i can. I'll be afk shortly. I will try and look at the other issue in a while.

**Comment by xexyl** [comment]
> When the [UPDATE 2 - ACTIONS REQUESTED from GH-issuecomment-2325407090](https://github.com/ioccc-src/mkiocccentry/issues/723#issuecomment-2325407090) are performed, we believe we are ready to close this issue.

Thanks! I missed that. I came to ask about it. Will close it as complete.

**Comment by xexyl** [comment]
> > When the [UPDATE 2 - ACTIONS REQUESTED from GH-issuecomment-2325407090](https://github.com/ioccc-src/mkiocccentry/issues/723#issuecomment-2325407090) are performed, we believe we are ready to close this issue.
> 
> Thanks! I missed that. I came to ask about it. Will close it as complete.

Actually ... is this premature? I'm not sure looking at the comment again.

How does it look on your end?

Also I want your thought on something else ...

# QUESTION

The JSON parser, if valid JSON, will print out:

```
valid JSON
```

I think that this might not be useful unless verbose option is used. Whether it would be a debug message or just `msg()` I don't know (what do you think?) but removing this text would be good. Though doing so would ruin my thought, maybe, that jparse subdirectory 'over there' need not be updated until after the next IOCCC.

Anyway let me know what you think on whether this issue can be considered complete (I'd also appreciate if you test that comment on your end .. and maybe I should try it again?) and if I should remove that message (and if so should it be shown via dbg() or msg() if `-v 1`). I can then go from there.

I'll be doing other things in the meantime.

**Comment by lcn2** [comment]
> Also I want your thought on something else ...
> 
> # QUESTION
> The JSON parser, if valid JSON, will print out:
> 
> ```
> valid JSON
> ```
> 
> I think that this might not be useful unless verbose option is used. Whether it would be a debug message or just `msg()` I don't know (what do you think?) but removing this text would be good. Though doing so would ruin my thought, maybe, that jparse subdirectory 'over there' need not be updated until after the next IOCCC.
> 
> Anyway let me know what you think on whether this issue can be considered complete (I'd also appreciate if you test that comment on your end .. and maybe I should try it again?) and if I should remove that message (and if so should it be shown via dbg() or msg() if `-v 1`). I can then go from there.
> 
> I'll be doing other things in the meantime.

# ANSWER

Please remove the "valid JSON" message, unless the verbosity is >= 1.

By default (and thus `-v 0`), the exit code from `jparse(1)` is sufficient.

**Comment by xexyl** [comment]
> > Also I want your thought on something else ...
> 
> > 
> 
> > # QUESTION
> 
> > The JSON parser, if valid JSON, will print out:
> 
> > 
> 
> > ```
> 
> > valid JSON
> 
> > ```
> 
> > 
> 
> > I think that this might not be useful unless verbose option is used. Whether it would be a debug message or just `msg()` I don't know (what do you think?) but removing this text would be good. Though doing so would ruin my thought, maybe, that jparse subdirectory 'over there' need not be updated until after the next IOCCC.
> 
> > 
> 
> > Anyway let me know what you think on whether this issue can be considered complete (I'd also appreciate if you test that comment on your end .. and maybe I should try it again?) and if I should remove that message (and if so should it be shown via dbg() or msg() if `-v 1`). I can then go from there.
> 
> > 
> 
> > I'll be doing other things in the meantime.
> 
> 
> 
> # ANSWER
> 
> 
> 
> Please remove the "valid JSON" message, unless the verbosity is >= 1.
> 
> 
> 
> By default (and thus `-v 0`), the exit code from `jparse(1)` is sufficient.
> 
> 
> 
> # UPDATE 0
> 
> 
> 
> Thank you in advance, @xexyl for the tool not printing 'valid JSON` unless there is some >=1 level of verbosity.

I was going to do it now but then I noticed that if it's invalid JSON it reports that as an error. 

To save time and effort how do you want that handled? Or you could open a pull request of course. 

I have to go get some food but I might have time to do this today if you get back to me soon enough. Otherwise I will be happy to do it tomorrow. 

Please advise what you prefer.

**Comment by xexyl** [comment]
Actually I just did it .. kept it as an error message as it's meant to be non-zero exit code. If you want it done differently please let me know.

I can sync it with the other repo or if you want you can do it after the merge of the changes. Must go for now though.

**Comment by lcn2** [comment]
> I was going to do it now but then I noticed that if it's invalid JSON it reports that as an error.
> 
> To save time and effort how do you want that handled? Or you could open a pull request of course.
> 
> I have to go get some food but I might have time to do this today if you get back to me soon enough. Otherwise I will be happy to do it tomorrow.
> 
> Please advise what you prefer.

There is an argument for issuing an error message if the JSON is invalid, even if there is no verbosity.  A `-v 0` does **NOT** mean, "no error message output", just perhaps no verbose error messages.

Perhaps if there is no verbosity (i.e., `-v 0`), and the JSON is invalid, the error message should simply be: "**invalid JSON**" written to stderr.

If JSON is invalid, and there is verbosity (i.e., >= 1) then the current level of extra error messages is warranted.

Thus this prints nothing and just does an `exit(0)`:

```sh
jparse good.json
```

This, prints the "**valid JSON**" message:

```sh
jparse -v 1 good.json
```

This writes a "**invalid JSON**" message to stderr:

```sh
jparse invalid.json
```

This, prints more detail errors:

```sh
jparse -v 1 invalid.json
```

That is what we would recommend.

**Comment by lcn2** [comment]
> That is true but the problem is the way it's included, the `../`. If we do the usual include (no directory) then we also require that users of the mkiocccentry toolkit have to install those libraries. But if we don't do that then we have the problem of how to clone this repo to the mkiocccentry repo?

The upper level sets `C_SPECIAL= -DINTERNAL_INCLUDE` when calling lower level Makefiles.  So the code for this repo should all be internal.

Recently we tested the "virgin build" by doing `make uninstall` for the dbg rep, the dyn_array repo, the jparse repo and this repo.  From that "virgin build" state, **NOTHING** was needed outside of the code in this repo: **NO** external include files and **NO** external libraries.  That was tested and verified for both macOS and Linux.  That was true for commit ba3da10061dfb9036d2a3216d220a7ad784f1295 from this repo as well as for the dbg repo, the dyn_alloc repo, and the jparse repo.  If that fails now, then a bug was re-introduced that needs to be fixed.

**Comment by xexyl** [comment]
> > I was going to do it now but then I noticed that if it's invalid JSON it reports that as an error.
> > To save time and effort how do you want that handled? Or you could open a pull request of course.
> > I have to go get some food but I might have time to do this today if you get back to me soon enough. Otherwise I will be happy to do it tomorrow.
> > Please advise what you prefer.
> 
> There is an argument for issuing an error message if the JSON is invalid, even if there is no verbosity. A `-v 0` does **NOT** mean, "no error message output", just perhaps no verbose error messages.
> 
> Perhaps if there is no verbosity (i.e., `-v 0`), and the JSON is invalid, the error message should simply be: "**invalid JSON**" written to stderr.
> 
> If JSON is invalid, and there is verbosity (i.e., >= 1) then the current level of extra error messages is warranted.
> 
> Thus this prints nothing and just does an `exit(0)`:
> 
> ```shell
> jparse good.json
> ```
> 
> This, prints the "**valid JSON**" message:
> 
> ```shell
> jparse -v 1 good.json
> ```
> 
> This writes a "**invalid JSON**" message to stderr:
> 
> ```shell
> jparse invalid.json
> ```
> 
> This, prints more detail errors:
> 
> ```shell
> jparse -v 1 invalid.json
> ```
> 
> That is what we would recommend, @xexyl.

I agree with this. But two things come to mind that have to be answered before this can be said to be complete (not this issue #2 but the error message issue).

First is that it uses the `err()` function. Should it just issue an error by `msg()` and then exit non-zero?

(ANSWER: no it should not as without the err() call it will just show a trap with the `not_reached()`.)

Second is this. We at first had warnings issued in the lever if a bad token was found. So for instance:

```sh
$ jparse -s '"'
Warning: jparse_lex: at line 1 column 1: invalid token: 0x22 = <">
syntax error in file - at line 1 column 1: <">
Warning: ./jparse: JSON parse tree is NULL
ERROR[1]: ./jparse: invalid JSON
```

I can't remember where exactly the `warning` line is in. This is unfortunate esp as...

```sh
$ git grep 'Warning' '*.[cly]'
json_util.c:1621:       fprint(stream, "\tWarning: JTYPE_UNSET: %s", json_type_name(node->type));
json_util.c:1800:                       fprint(stream, "\t{%s%s}: Warning: wrong name type: %s",
json_util.c:1848:                   fprstr(stream, "\tWarning: set == NULL");
json_util.c:1851:                   fprstr(stream, "\tWarning: s == NULL");
json_util.c:1887:                   fprstr(stream, "\tWarning: set == NULL");
json_util.c:1890:                   fprstr(stream, "\tWarning: s == NULL");
json_util.c:1926:                   fprstr(stream, "\tWarning: set == NULL");
json_util.c:1929:                   fprstr(stream, "\tWarning: s == NULL");
```

However I do guess it's in `yyerror()` in `jparse.y`. But where does the string `Warning:` come from? Okay we might want to check the 'invalid token' line. Let's see.... I do remember that and adding the character.

Oh! Of course. That comes from the `warn()` function! So.... it's in `jparse.l`. 

We can change the line:

```
                            warn(__func__, "at line %d column %d: invalid token: 0x%02x = <%c>", yylloc->first_line, yylloc->first_column, *yytext, *yytext);
```

to use `err()`. Or so we would think. Look:


```c
                            err(1, __func__, "at line %d column %d: invalid token: 0x%02x = <%c>", yylloc->first_line, yylloc->first_column, *yytext, *yytext);
```

but that then gives:

```
ERROR[1]: jparse_lex: at line 1 column 1: invalid token: 0x22 = <">
```

which is not nearly as helpful and it does not let `jparse` show more details. This is what I'm going to commit shortly and then I'll sync to the other repo (after also committing the removal of the timestamp in Makefile):

```
syntax error in file - at line 1 column 1: <">
Warning: ./jparse: JSON parse tree is NULL
ERROR[1]: ./jparse: invalid JSON
```

without `-v 1` but with `-v 1`:

```
debug[1]: at line 1 column 1: invalid token: 0x22 = <">
syntax error in file - at line 1 column 1: <">
Warning: ./jparse: JSON parse tree is NULL
ERROR[1]: ./jparse: invalid JSON
```

That looks much nicer I think since the debug message shows the token hex value and the second line (without debug output) shows that there is a syntax error and at what line and column along with just the character itself.

The warning line with the parse tree being NULL seems to be good esp as it might not always be because of invalid JSON (unlikely as that seems): it also comes from `jparse` itself, not the parser. The `syntax error` does come from the parser.

After these things I will look at the 'virgin build' again and then if all is good we can consider this completely resolved. I think that'll be fine.

**Comment by xexyl** [comment]
The error reporting has now been improved and the Makefile timestamps have been removed! I will sync to the other repo in a bit.

I will test the 'virgin build' sometime soon too. If all is good I can look at the mkiocccentry repo comment(s) and then try and address them (maybe working on it some). After that I have other things I must do. If I have time after that I will end up working more on the mkiocccentry repo.

But now the error messages/reporting in the jparse tool is much better.

**Comment by xexyl** [comment]
Annoying typing `#n` links to this. I have to remember to use inline code blocks or not use the `#`. More annoying is that once I deleted the comment to fix it still keeps it. Oh well. I should remember that now .. I thought there was a way to not link to it but maybe not.
